### ISSUE TITLE
Sudden flip of the sign in velocities

### ISSUE THREAD

**Issue by jhprinz** [choderalab/openpathsampling#416]
@bolhuis mentioned this in an example. That suddenly in a trajectory there is a flip of sign which should not happen.

This needs to be fixed!!!

**Comment by jhprinz** [comment]
Identified the issue. Just looking for an elegant solution. This is highly non-trivial and involves
1. features
2. memory efficient caching
3. Snapshot classes
4. Paired creation of Snapshots

need to think about this a little. Try to post a solution soon.

There is one design question I need to know your opinion:
Storing reversed snapshots have a flag. Do we always want to store the same velocities in the database and determine flip from the reversed bool. Or (like in the case of ToySnapshots) do we allow that velocities and *1.0 * velocities is saved as a pair. That is nicer when looking at the storage without OPS.

Remember in MDSnapshots we store two times a reference to the same velocities and only use the flag.

**Comment by dwhswenson** [comment]
It seems to me that it is pretty useful to only save a flag for reversal -- much smaller files. Only one copy of the velocities is saved, and it is flipped programmatically if the snapshot has its `is_reversed` flag set. So I think we should do that. I didn't realize the `ToySnapshot`s were saving both versions of velocity -- of course, for the 2D systems, that isn't a big hit on storage space. But for 10k atoms, it would be a problem.

As for looking at snapshots outside of OPS, I guess the question is how will people do that. If you're using these objects, but never using any kind of reversal, than the `is_reversed` flag is just an extra (unused) thing in storage. I don't think that's a problem. The only confusion would come from either (a) trying to work directly from the netCDF file without going through netCDF+'s object re-building process, of (b) trying to copy trajectories generated from a code with the `reversed` flag to a code that doesn't use it (and probably then using direct netCDF or some other way to analyze it).

(a) I don't think many people will work directly from the netcdf file -- most will probably go through the NetCDF+ interface, which means that they will work from the `Snapshot` objects (especially if we plan to split the engines/snapshots/trajectories into a reusable package later). As long as the things stored in the netCDF are well-documented, a user can write another reader for that (e.g., in C instead of Python) to re-create the trajectories correctly.

(b) Pretty much the same as with (a), although I'll add a suggestion of a `Snapshot.copy_clean(snap)` function that returns a deep copy of the snapshot with time set to forward (`is_reversed=False`), no matter what the input snapshot. Similarly, something like `Trajectory.copy_clean(traj)` which does a deep copy of the whole trajectory would be good. That way, a user can always convert the snapshots into a "forward-time" version that will be easily compatible with other things -- this would mainly be used as a step before exporting trajectories to a different program. But even that seems like an extra safety precaution that might only be necessary in obscure edge cases.

**Comment by dwhswenson** [comment]
While you're working on snapshots, another thought: Do we need the `Configuration` and `Momentum` objects? Can't a snapshot just have additional features of `box_vectors` and the various energies?

I'd like to re-use some of this for another project I'm playing with, but in that case I need to save ACTUAL momenta. It would be nice to get rid of the poorly-named `Momentum` class, Now that you've created snapshot features, I don't think we need it any more.

**Comment by jhprinz** [comment]
Let me address this altogether 

Well, storing only the flag is what I wanted in the first place. The reason that I stored always pairs (which in the case of ToySnapshot lead to double saving) is that the CVs are effectively also snapshot variables and have the same size. So, since a CV can have completely different values for a snapshot and its reversed we needed to save both of them. At the time it looked cleaner to me to just have one row on the table per snapshot and not per snapshot/reversed pair. Thus double the space in the file.
For MDSnapshots this is fine since so far the store only a pointer to a Congifuration or Momentum object. Hence no problem. For ToySnapshots with lots of atoms or dimensions this will be a big overhead. 

My new proposal is this:
1. Store only once for each snapshot/reversed pair. This will reduce diskaccess and storage
2. CVs that are time_reversible do only need minor changes
3. CVs that are _not_ time_reversible will treated as 2 CVs internally. One for the reversed=False and one CV for the reversed=True. This should keep efforts for implementation minimal
4. In memory I would keep the pairs that link to each other (unless I find an elegant way to use a proxy ReversedSnapshot)
5. The reason for the configuration / momentum snapshots is partially legacy and partially storage related. I also found the hierarchy nice that you have kinetic and potential part separated. The advantage is that you can store Snapshots without configuration and/or without momentum and we can reuse the same coordinates for the same Snapshot. Like storing 10 Snapshots with different momenta only. If we do not care about this. Then we can save this as ordinary features. 
6. We could have lazy loading features, but so far only if they are storable objects, Numpy array do not work. So when storing MDSnapshots using simple features only you will read the whole snapshot not just coordinates. I do not think this is a problem.
7. IMPORTANT! It would simplify storage a lot if we say that all Snapshot attributes (features) in the reversed Snapshot are computable from the non-reversed version (and vice verse). This would mean
that storage can _always_ store only one side and infer the other. CVs will still allow to store informations that depend in reversal. My idea is to remove the user-implementation of the feature and let storage to all of that efficiently. The user would prepare prepare a template Snapshot and the storage would do the rest. I will make a sample PR so explain this.

**Comment by jhprinz** [comment]
I forgot an important point. When you save a trajectory then you only save references to a snapshot IDX. And this IDX need to contain information about reversed or not. Hence a frame and its reversed need different IDXs. So `Configuration` and `Momentum` are meant to not store arrays of data more than once.

We could make the Configuration and Momentum object simpler and remove box_vector and energies form it and then it is only a numpy array. So use caching etc we still need the wrapper object. Example
```
coords = my_numpy_array_with_coordinates # no worries about units :)
storage.coordinates.save(coords)

# the next command will store again since np arrays are not hashable and
# we cannot figure out if it has been stored!
storage.coordinates.save(coords)
```

I think this needs way more thinking. So revert last comment. I will keep it the way it is so far. If you want to use ToyDynamics with large sets of atoms we should provide an ExtendedToySnapshot what can save.

One last idea that I just had is to use a simple abstractSnapshot like so
```
Snapshot(
    is_reversed=...,
    frame=...
)
```
The `Frame` contains all the data that should be stored in whatever form. is_reversed is also hidden since the user does not need it. All you have to implement are getters for the Snapshot to convert from 
`frame.something` to the Snapshot properties. This could clear it up and allow better naming that Momentum and Configuration!

**Comment by dwhswenson** [comment]
> So revert last comment. I will keep it the way it is so far. 

This statement is confusing. Keep *what* the way it is so far? We have a bug that needs to be fixed, so obviously you're planning to change something -- do you just mean that I should ignore the "new proposal" part of in [your earlier comment](https://github.com/choderalab/openpathsampling/issues/416#issuecomment-172519640)? (For now, I'm assuming that's what you meant.)

> One last idea that I just had is to use a simple abstractSnapshot like so
> 
>     Snapshot(
>         is_reversed=...,
>         frame=...
>     )
>
> The `Frame` contains all the data that should be stored in whatever form. `is_reversed` is also hidden since the user does not need it. All you have to implement are getters for the Snapshot to convert from `frame.something` to the Snapshot properties.

If we can avoid getters, I would prefer that. This is inner-loop stuff, so I don't think it is premature to consider function-call overhead.

***

My thoughts:

#### Big picture

If I'm understanding correctly, the fundamental problem is that `Snapshot` uses an intermediate container object to access the actual attributes, whereas `ToySnapshot` does not. Is this what leads to the bug that Peter observed? The solution is, of course, to have all types of snapshots behave the same in this regard. (Source of bug or not, we should do this. It would be more straightforward.)

#### `Configuration` and `Momentum` renaming?

These are container objects. They can be named explicitly as such, which I think would be clearer. `ConfigurationContainer` and `KineticContainer` would be my recommendations (if we keep these objects at all; see following).

#### Grouped containers or atomic containers

Right now, `Configuration` includes coordinates, box vector, and potential energy. Different engines may have different combinations of these. Rather than having a different container object for each likely combination (and trying to come up with reasonable names for them), I think it would be better to have each as a separate feature -- a more "atomic" feature. A given `Snapshot` subclass will know what to do with each of its `__features__` on time-reversal.

However, that depends on whether we actually need container objects in addition to features.

#### Containers or direct?

If there's a way to use direct access, instead of containers, to get these attributes, that would be better. That's faster, easier to understand, and simpler to implement for new snapshot features.

If I understand correctly, the only reason for using a container object is to make it cacheable (and possibly otherwise compatible with netCDF+). Would be be possible to subclass `numpy.ndarray` instead? There are a few tricks to doing that correctly, but [the documentation on it is pretty clear](http://docs.scipy.org/doc/numpy-1.10.1/user/basics.subclassing.html). (I used those docs before when making `LookupFunction` compatible with numpy ufuncs.) This should lead to faster code than a wrapper class, which has an extra layer of indirection to get access the data.

#### Summary

I think the things we care about are:
* Fast access to the actual data when snapshot is already in memory
* Ability to work with caching and storage in netCDF+
* Reusability for other projects
* Uniform treatment of features across all types of snapshots

Maybe I'm missing caveats that prevent this, but my preference would be, in order:

1. Snapshot features shouldn't require containers (but perhaps require a special `CacheableArray` subclass of `numpy.ndarray` -- which would be a part of netCDF+, since it would be useful anywhere that we use the caching/storage mechanisms of netCDF+).

2. If that subclass isn't possible, then each feature should have a separate container object, which only holds one `np.array` (or one float).

3. If none of the rest of this happens, we should at least rename the container objects as `[Something]Container`, to make it clear what they are. This gets rid of the `Momentum` object name, which has always bothered me.

Is there anything that isn't possible in approach 1? I suppose you would have to wrap single-float-features (e.g., energies) to 1D arrays, but you could even have a special subclass that handles that pretty easily. These subclasses could even be pretty transparent to the user: on creation, the snapshot internally turns everything into `netcdfplus.CacheableArray` (or whatever you want to call it), but the user never has to directly deal with that class, and only sees it as an `np.array`.

***

Addendum -- why I'm suddenly thinking so much about speed:

We've talked a lot about re-using some parts of OPS for other projects, and I've mentioned that I want to create a Python-based interface for my semiclassical work. Well: 

http://github.com/dynamiq-md/dynamiq_engine

**Comment by jhprinz** [comment]
Some more food for thought and later discussion... I think a have almost found a good way to make this more simple and still fast...

# Storage of Snapshots and CVs

Sorry my last post was done in a bit of a hurry. Let me add a few points to your post to make the problem a little clearer.

## The bug

The problem was partially due to the difference you describe. It was only relevant for direct stored values and only for values that switch momentum and only if I used special loading that uses preloaded reversed ones... LongStory short. The quick fix is almost trivial, but I thought since there were more points to address we could clarify these and get things done properly.

### Important Aspects

There are several aspects that I consider important when storing snapshots and you mentioned most of them before:

1. Object complexity : Simple is better, so I would prefer direct attributes and not wrapped ones. This includes Reusability and uniform access of attributes.
2. Loading speed : It is affected by loading/creating in pairs or single. If objects are loaded partially or full or even only by reference.
3. Saving speed : This is usually neglegiable since it is only done once. 
4. Memory efficiency : Paired or Single loading into cache. It is also important if attributes are very large and need to efficiently loaded into memory. This is where caching is important. 
5. Storage efficiency : paired or single storage. Affects storage of CVs as well
6. Creation speed : Good if objects are simple and do not reference complex object
7. Referenceable : Each possible snapshot needs to be referenceable to allow CVs to work

### Split between memory and storage representation

I think we need to figure out two separate parts. 

1. The memory representation of `Snapshot` so that it supports features. 
2. Storage of `Snapshot` so that is supports features

Right now 

### Double storage vs Compatibility

We need to reference snapshots like in trajectories and if possible I want to do that by a single int and not a more complex object. If so, then I think indexing in pairs makes sense. So `0` references the first configuration, `1` references the first configuration reversed, `2` references the second configuration etc. This part is clear I think and makes sense. 

Do we also want to reflect this numbering in the storage? Yes or No. This will double the storage for snapshots but will be easier when someone looks at the file without OPS. If we do not want to use that doubling in storage we will reuse the current saving but devide `snapshot_idx / 2` to get the actual idx in tht storage. 

I tried this once but did not use this because of CVs. Now I think this would be the fastest and most memory efficient storage with the caveat the referencing number do not directly point to the position in storage.

To handle CVs that are not time_reversible we need to change the storage of CVs so that it will use a second table for the reversed indices. Again this will save us from storing reversible CVs twice.

- My choice: Use single storage

All of this is independent of how we store a single snapshot

### Inferring the reversed snapshot

If we do not store the reversed copy we need a generic way to create the memory representation of the reversed snapshot. It is crucial for CV and Snapshot caching that there is actually a hashable memory object the represents the reversed object. I see 2 choices:

1. Create a real copy that is independent of the original snapshot. 

2. Create a proxy copy that references the original and reuses as much as possible

The second is more complicated but faster and more memory efficient. No new objects have to be created. 

All of this is independent on the feature implementation and if some features are loaded lazy or not. In 1. or 2. we could keep the lazy loading. 

Another question is if we wish to 

1. create the pair on loading or 
2. create the reversed snapshot on request

I prefer the 2. which should be possible. It is also faster since in most cases we need only one copy.

Next to note is that reversibility is idempotent. Means we can just apply the reversal operation and store the pair of original and reversed copy with opposite assignment of which is the original. So the flag `is_reversed` is a matter of perspective.

In principle it means that we could remove the necessity of the flag completely. It would be replaced by the `ReversedProxy`

We will also store the version that is first asked to be stored. What does that mean? If you say `storage.save(snap.reversed)`. It will check if `snap.reversed` is already saved, that is done by the caching automatically. If not it checks if `snap.reversed.reversed` is saved and do nothing and only return the correct storage idx (which differs from the idx of `snap` by one). If none of them is stored it will store `snap.reversed` as the original and return the new idx. 

It also means that loading the object is independent of reversed state, because reversed or not is only determined either by requestion the reversed snapshot or loading reversed copy. In any case we start we an unreversed copy and reverse it.

Consequence for features is that a user can define the loading and saving independent of worrying about reversed or not. It also means that a user creating a new snapshot class needs to implement the mechanism that returns reversed values. Usually only necessary for velocities.

This removal of double storage of snapshots might make the use of numpy array containers unnecessary. All we would loose is to store a configuration with several velocities. Like it happens using two-way shooting. You would have to store the configuration twice. 

The use of containers is however optional. We could have a feature with containers and one without.


### Warp numpy arrays

I would avoid the ndarray subclass. I think this could get complicated. The problem might (not sure) be that the objects needs to be hashable which we do using `id() / 8` as is the python default. I am not familiar with memory handling and reusing of numpy so that is where we could run into trouble. Another point would be handling units but less important. 

The current approach was a compromise to a have few objects that gather attributes most-likely being used together. 


### How the snapshot structure could look like

```py
Snapshot()
    reversed : Snapshot()
        lazy pointer to the reversed snapshot
    engine : Engine()
        the engine that was used to generate the snapshot
    coordinates : numpy.ndarray, shape=(atoms, spatial)
        pointer to `.data.coordinates`
    velocities  : numpy.ndarray, shape=(atoms, spatial)
        pointer to `.data.coordinates`
    box_vector : numpy.ndarray, shape=(spatial,spatial)
        pointer to `.data.box_vectors`
```

energies will be turned into a CV and needs to be accessed using

```
ke = storage.cvs['kinetic_energies']
print ke(my_snapshot)
```

### Handling of features

Currently features only implement code to save. I would like to make a feature a bridge between the Snapshot and the Storage so that it implements parts of both. An example

```py
"""
coordinates : simtk.unit.Quantity wrapping Nx3 np array of dimension length
    atomic coordinates (default: None)
"""
attributes = ['velocities']

# the feature specific initialization
def store_init(store):
	store.create_variable('velocities', 'numpy.float32',
	    dimensions=('atom', 'spatial'),
	    chunksizes=(1, 'atom', 'spatial')
	    )

# how to get the time-reversed feature
reversible_attribute('velocities', lambda x : -1.0 * x)
```

Based on the code in the feature the appropriate tables in the netCDF file are created and the store _learns_ how to create reversed copies if necessary.
If `store_init` is not specified then it will infer the structure from the `template.velocities` content if possible. Default for `reversed` it the identity `return features`.

In a more complex situation (with a wrapper as before) it could look like this

```py
# which attributes will be added and stored
attributes = ['momentum']
# which properties will be added (and computed)
properties = ['velocities', 'kinetic_energy']

# the netcdf+ specific initialization
def netcdfplus_init(store):
    store.storage.create_store('momenta', MomentumStore())

    store.create_variable(
    	'momentum', 
		'lazyobj.momenta'
	 )

# implement the properties that will be added to the snapshot
def velocities(snapshot):
    """
    velocities : simtk.unit.Quantity wrapping Nx3 np array of dimension length
        atomic velocities (default: None)
    """
	return snapshot.momentum.velocities
	
reversible_attribute('velocities', lambda x : -1.0 * x)

def kinetic_energy(snapshot)
    """
	kinetic_energy : simtk.unit.Quantity of units energy/mole
        kinetic energy
    """
	return snapshot.momentum.kinetic_energy

# Implementation of the MomentumStore
class MomentumStore(ObjectStore):
	...
    
```

Finally we use a decorator to add features. This is cleaner than a special attribute. Even better would be the use of a metaclass, but in this case we need to remove the use of abstract base class since it also uses a metaclass.

Example

```py
@set_features(
    features.momentum,
    features.configuration
)
class MDSnapshot(AbstractSnapshot):
	pass
```

All of this is like a nicely configurable factory of Snapshots. You give the features and it will add everything to the class that is needed. In most cases you do not have to write any code. And it should still be fast. The factory could also take care of the docstrings and add these appropriately to features.

**Comment by dwhswenson** [comment]
I'm still very concerned about speed, especially on small systems.

Maybe the internal `data` construct can be part of that -- that could be a non-Python (i.e., Cython or C extension) for speed. Keep in mind that, when doing small systems (which is much of what I do) the bottleneck right now will probably be in Snapshot creation. Memory allocation is expensive.

In many cases, I don't actually care to save all the trajectories (if I'm doing 50 million trajectories of a 25 atom system, for example), so I'd rather only allocate once and re-use that memory, only copying a few trajectories to storage. This would mean that `Snapshot`s (or some subclass of `AbstractSnapshot`) can have mutable coordinates. This allows you to, for example, create a full trajectory and pass it through some averaging (e.g., to calculate the average of some CV as a function of time for many fixed-length trajectories -- which is really all that the semiclassical initial value representation does.)

Even if we can't do that in OPS because of weirdness with tracking the shooting point, that should be completely possible in non-shooting approaches to generating many trajectories. (Again, my goal here is to re-use this code for other projects.)

A few of your specific headings:

#### Important Aspects

In addition to thinking about perturbative path sampling, if we're planning for this to be a reusable module, we should consider other (more common) approaches. That does change the analysis at a few points.

> Object complexity : Simple is better, so I would prefer direct attributes and not wrapped ones. This includes Reusability and uniform access of attributes.

We're agreed on this.

> Loading speed : It is affected by loading/creating in pairs or single. If objects are loaded partially or full or even only by reference.

I think you've already done a lot of clever things to manage this.

> Saving speed : This is usually neglegiable since it is only done once.

Not always true. We already had to reduce the frequency of saving in the ipynb tests because the toy models were disk-bound. In small systems, you can generate trajectories more quickly than you can save them to disk, and probably more quickly than you can move them to off-cache RAM.

> Memory efficiency : Paired or Single loading into cache. It is also important if attributes are very large and need to efficiently loaded into memory. This is where caching is important.

Yes, although I think loading by reference will help a lot here.

> Storage efficiency : paired or single storage. Affects storage of CVs as well

This is *extremely* important. Output from a large TIS run will be TB of data. Double, or even 50% increase, in storage can mean needing to buy another hard drive.

> Creation speed : Good if objects are simple and do not reference complex object

Not necessarily. This is frequently going to be a bottleneck in small systems. Memory allocation is expensive.

> Referenceable : Each possible snapshot needs to be referenceable to allow CVs to work

We should see if there's a way to loosen this. Again, I'm thinking about avoiding the need to re-allocate snapshots.

#### Split between memory and storage representation

Looks like this section was incomplete when you posted. But yes, these are two separate topics to handle. 

#### Double storage vs Compatibility

Some of you wording here is unclear: do you mean `Snapshot` when you say "configuration"? ("reversed configuration" means nothing.)

This heading seems closely tied to the next one, and any reasons for making a decision are based on the next one. So....

#### Inferring the reversed snapshot

> I see 2 choices:
> 1. Create a real copy that is independent of the original snapshot.
> 2. Create a proxy copy that references the original and reuses as much as possible
> The second is more complicated but faster and more memory efficient. No new objects have to be created.

Unclear to me: do you mean that the proxy copy is not actually an instance of `Snapshot`? I thought it had to be (and therefore ONE new object is created). But I think we're in agreement that we don't want a new copy of the velocities.

> Another question is if we wish to
> 1. create the pair on loading or
> 2. create the reversed snapshot on request
> I prefer the 2. which should be possible. It is also faster since in most cases we need only one copy.

Agreed. Especially in light of the fact that *most* situations (outside of path sampling) have no need for the reversed copies. This is also what determines the answer to the previous topic: there's no sense in saving alternating rows of forward and reversed snapshots if you only might have forward snapshots.

> It also means that loading the object is independent of reversed state, because reversed or not is only determined either by requestion the reversed snapshot or loading reversed copy. In any case we start we an unreversed copy and reverse it.

I'm confused. How will a saved trajectory know whether its snapshots are `reversed` or not? If direction isn't associated with a saved snapshot, do you need to save a list of directions with each `Trajectory`?

> This removal of double storage of snapshots might make the use of numpy array containers unnecessary. All we would loose is to store a configuration with several velocities. Like it happens using two-way shooting. You would have to store the configuration twice.

Order of 1% increase in disk space, perhaps less. That's still a 10GB per TB, so not nothing, but I'm willing on that.

#### Wrap numpy arrays

> I am not familiar with memory handling and reusing of numpy so that is where we could run into trouble. Another point would be handling units but less important.

None of this should be a problem. Numpy arrays are expected to be subclassed (as in those docs). But wrapping can also work.

> The current approach was a compromise to a have few objects that gather attributes most-likely being used together.

Why? I still don't see the value in the `Configuration` and `Momentum` objects. It makes more sense to me to wrap each snapshot feature independently. It also solves confusion around naming.

#### How the snapshot structure could look like

From the user perspective, that looks good to me.

#### Handling of features

> Currently features only implement code to save. I would like to make a feature a bridge between the Snapshot and the Storage so that it implements parts of both.

This makes it more simple at the cost of making it harder to speed up. We should talk about this. I think speed might be more important here.

... and sending this now, and you'll probably not have time to read it over before we Skype. Sorry about that! :-/

**Comment by jhprinz** [comment]
I think #429 solved this one. Can we close it? Or do we need some of this as ref docs?

**Comment by dwhswenson** [comment]
Do we have something in nosetests for this? This is a very dangerous regression, and we need to be sure there's a test in place that will alert us if this goes wrong. We should not close until that it done.
